src/Course/FileIO.hs

```diff
@@ -81,50 +82,57 @@ the contents of c
 
 -- Given the file name, and file contents, print them.
 -- Use @putStrLn@.
-printFile ::
-  FilePath
-  -> Chars
-  -> IO ()
-printFile =
-  error "todo: Course.FileIO#printFile"
+printFile :: FilePath -> Chars -> IO ()
+printFile path chars = putStrLn ("The path: " ++  path ++ ". The chars: " ++ chars)
 
 -- Given a list of (file name and file contents), print each.
 -- Use @printFile@.
-printFiles ::
-  List (FilePath, Chars)
-  -> IO ()
-printFiles =
-  error "todo: Course.FileIO#printFiles"
+printFiles :: List (FilePath, Chars) -> IO ()
+printFiles = voidSequenceMap $ uncurry printFile
 
 -- Given a file name, return (file name and file contents).
 -- Use @readFile@.
-getFile ::
-  FilePath
-  -> IO (FilePath, Chars)
-getFile =
-  error "todo: Course.FileIO#getFile"
+getFile :: FilePath -> IO (FilePath, Chars)
+getFile path = readFile path >>= \content -> return (path, content)
+-- getFile path = do
+--   content <- readFile path
+--   return (path, content)
+
 
 -- Given a list of file names, return list of (file name and file contents).
 -- Use @getFile@.
-getFiles ::
-  List FilePath
-  -> IO (List (FilePath, Chars))
-getFiles =
-  error "todo: Course.FileIO#getFiles"
+getFiles :: List FilePath -> IO (List (FilePath, Chars))
+getFiles = sequenceMap getFile
 
 -- Given a file name, read it and for each line in that file, read and print contents of each.
 -- Use @getFiles@ and @printFiles@.
-run ::
-  FilePath
-  -> IO ()
-run =
-  error "todo: Course.FileIO#run"
+run :: Optional FilePath -> IO ()
+run Empty = putStrLn "Something went wrong"
+run (Full path) = readFile path >>=
+  \file -> getFiles (lines file) >>=
+  \files -> printFiles files
+-- run path = do
+--   file <- readFile path
+--   let content = lines file
+--   files <- getFiles content
+--   printFiles files
+
+sequenceMap :: Monad m => (a -> m b) -> List a -> m (List b)
+sequenceMap f = sequence . (<$>) f
+
+voidSequenceMap :: Monad m => (a -> m b) -> List a -> m ()
+voidSequenceMap f = void . sequenceMap f
+
+optionalPath :: List FilePath -> IO (Optional FilePath)
```

Comment:
maybe:

```
-- could have sworn this function was in the List module, but it doesn't appear to be — it's in Data.Maybe in normal haskell as `listToMaybe` tho with the `Maybe` data type.
optionalHead :: List a -> Optional a 
optionalHead (x :. _) = Full x
optionalHead _ = Empty

optionalPath :: List FilePath -> IO (Optional FilePath)
optionalPath = pure . optionalHead
```

Reply:
we both called it optionalHead too, hehe